appcelerator/titanium_mobile
# fix(android): improve Ti.Geolocation API reliability in 9.0.0

**JIRA:**
https://jira.appcelerator.org/browse/TIMOB-27739

**Summary:**
- As of Titanium 9.0.0, "device" or "production" builds would no longer auto-add below permissions to "AndroidManifest.xml" if JS used `Ti.Geolocation` APIs.
  * `android.permission.ACCESS_COARSE_LOCATION `
  * `android.permission.ACCESS_FINE_LOCATION `
- The `Ti.Geolocation` APIs would never work if `play-services-base` library was included and `play-services-location` library was missing.
  * Added check for classes in both libraries before using Fused Location feature.
  * Can be an issue in 9.0.0 since app devs can set up their own custom `build.gradle` file.

**Test:**
1. Create a Classic app project.
2. Add the below "app.js" code to project.
3. Add the below "build.gradle" file to project.
4. Build and run on an Android device. (Not the emulator.)
5. Verify that app prompts you for location permission. Tap the "Allow" button.
6. Verify app fetches location data and shows it onscreen.

./platform/android/build.gradle
```groovy
dependencies {
	implementation 'com.google.android.gms:play-services-base:17.1.0'
}
```

./Resources/app.js
```javascript
function monitorLocation() {
	function start() {
		Ti.Geolocation.addEventListener("location", function(e) {
			label.text = JSON.stringify(e, null, 4);
			Ti.API.info("@@@ location received: " + JSON.stringify(e));
		});
	}
	var hasPermission = Ti.Geolocation.hasLocationPermissions(Ti.Geolocation.AUTHORIZATION_WHEN_IN_USE);
	if (!hasPermission) {
		Ti.Geolocation.requestLocationPermissions(Ti.Geolocation.AUTHORIZATION_WHEN_IN_USE, function(e) {
			Ti.API.info("@@@ permission result: " + JSON.stringify(e));
			if (e.success) {
				start();
			} else {
				label.text = "\nFailed to acquire location permission from end-user.";
			}
		});
	} else {
		start();
	}
}
var window = Ti.UI.createWindow();
var scrollView = Ti.UI.createScrollView({
	layout: "vertical",
	scrollType: "vertical",
	width: Ti.UI.FILL,
	height: Ti.UI.FILL,
});
var label = Ti.UI.createLabel({
	text: "\nWaiting for GPS data...",
	top: (Ti.App.iOS ? "25dp" : "5dp"),
	height: Ti.UI.SIZE,
	left: "5dp",
	right: "5dp",
	bottom: "5dp",
});
scrollView.add(label);
window.add(scrollView);
window.addEventListener("open", function(e) {
	monitorLocation();
});
window.open();
```
